phpBB/phpbb/db/migration/schema_generator.php

```diff
@@ -84,171 +98,239 @@ public function get_schema()
 			}
 
-			// Only run this check after the first run
-			if ($check_dependencies)
+			$split_position = strpos($change_type, '_');
+			$schema_change_type = substr($change_type, 0, $split_position);
+			$schema_type = substr($change_type, $split_position + 1);
+
+			$action = null;
+			switch ($schema_change_type)
 			{
-				$this->check_dependencies();
-				$check_dependencies = false;
+				case 'add':
+				case 'change':
+					$action = function(&$value, $changes, $value_transform = null) {
+						self::set_all($value, $changes, $value_transform);
+					};
+				break;
+				case 'drop':
+					$action = function(&$value, $changes, $value_transform = null) {
+						self::unset_all($value, $changes);
+					};
+				break;
+				default:
+					throw new \UnexpectedValueException("$migration_class contains undefined schema changes: $change_type.");
 			}
-		}
 
-		ksort($this->tables);
-		return $this->tables;
+			switch ($schema_type)
+			{
+				case 'tables':
+					$action($this->tables, $changes);
+				break;
+				default:
+					$this->for_each_table(
+						$changes,
+						$action,
+						$column_map[$change_type],
+						self::get_value_transform($schema_change_type, $schema_type)
+					);
+			}
+		}
 	}
 
 	/**
-	* Check if one of the migrations files' dependencies can't be resolved
-	* by the supplied list of migrations
-	*
-	* @throws \UnexpectedValueException If a dependency can't be resolved
-	*/
-	protected function check_dependencies()
+	 * Apply `$callback` to each table specified in `$data`.
+	 *
+	 * @param array			$data				Array describing the schema changes.
+	 * @param callable		$callback			Callback function to be applied.
+	 * @param string|null	$column				Column of the `$this->tables` array for the table on which
+	 * 											the change will be made or null.
+	 * @param callable|null	$value_transform	Value transformation callback function or null.
+	 */
+	private function for_each_table(array $data, callable $callback, $column = null, $value_transform = null)
 	{
-		// Strip duplicate values from array
-		$this->dependencies = array_unique($this->dependencies);
+		foreach ($data as $table => $values)
+		{
+			$target = &$this->tables[$table];
+			if ($column !== null)
+			{
+				$target = &$target[$column];
+			}
 
-		foreach ($this->dependencies as $dependency)
+			$callback($target, $values, $value_transform);
+		}
+	}
+
+	/**
+	 * Set an array of key-value pairs in the schema.
+	 *
+	 * @param mixed			$schema				Reference to the schema entry.
+	 * @param mixed			$data				Array of values to be set.
+	 * @param callable|null $value_transform	Callback to transform the value being set.
+	 */
+	private static function set_all(&$schema, $data, $value_transform = null)
+	{
+		$data = (!is_array($data)) ? [$data] : $data;
+		foreach ($data as $key => $change)
 		{
-			if (!in_array($dependency, $this->class_names))
+			if ($value_transform !== null)
```

Comment:
You think it could be safer to also check for is_callable here?

Reply:
Sure, all just type hints should make it more robust.